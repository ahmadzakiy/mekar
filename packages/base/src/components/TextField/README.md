# Mekar Text Field

## Installation
Install the component via NPM

```bash
yarn add mekar-text-field
```

## How to Use

### Import as Global Component
With this kind of import, you will include the whole bundle of the library, but you can use all components directly in Vue Instance.

```js
import { createApp } from 'vue'
import App from './App.vue'

// components
import TextField from 'mekar-text-field'

const app = createApp(App)
app.use(TextField)
app.mount('#app')
```

### Default Import
With this kind of import, you will only include component that've been imported.

```js
// components
import TextField from 'mekar-text-field'
```

## Contribution
Everyone is welcome to contribute to Mekar Library
