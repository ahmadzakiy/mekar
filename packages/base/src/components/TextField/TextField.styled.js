import styled from "vue3-styled-components";

const inputProps = {
  size: String,
  isDisabled: Boolean,
  isError: Boolean,
  isFocus: Boolean
}

const StyledInput = styled('input', inputProps)`
  position: relative;
  box-sizing: border-box;
  -webkit-appearance: none;
  -moz-appearance: none;
  outline: none;
  box-shadow: none;
  letter-spacing: normal;
  color: #3d3d3d;
  font-family: sans-serif;
  font-size: 16px;
  line-height: 20px;
  width: 100%;
  height: ${props => props.size === 'small' ? '36px' : '44px' };
  padding:${props => props.size === 'small' ? '0 12px' : '0 20px' };
  background: #ffffff;
  /*border-radius: 8px;*/
  border: 2px solid #d6d6d6;
  transition: all .2s ease-in-out;

  &:hover {
    border-color: #2572b7;
  }

  &::placeholder {
    color: #8d8d8d;
  }

  ${props => props.isFocus && `
    border-color: #2572b7;
  `}

  ${props => props.isError && `
    border-color: #ec0c0c;

    &:hover {
      border-color: #ec0c0c;
    }
  `}

  ${props => props.isDisabled && `
    cursor: not-allowed;
    background-color: #f2f2f2;
    border-color: #f2f2f2;
    color: #b6b6b6;

    &:hover {
      border-color: #f2f2f2;
    }
  `}
`

const StyledLabel = styled.label`
  display: block;
  color: #3d3d3d;
  font-family: sans-serif;
  font-size: 14px;
  line-height: 20px;
  margin-bottom: 8px;
`

export {
  StyledInput,
  StyledLabel
}
