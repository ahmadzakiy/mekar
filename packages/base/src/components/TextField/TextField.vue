<template>
  <label
    class="label"
    :for="id"
  >
    {{ label }}
  </label>
  <input
    v-bind="$attrs"
    class="input"
    :class="classes"
    :name="id"
    :value="modelValue"
    :disabled="isDisabled"
    @input="handleInput"
    @focus="handleFocus"
    @blur="handleBlur"
    @click="handleClick"
  >
</template>

<script>
  import { SIZE } from './type'

  export default {
    name: 'TextField',
    version: '1.0.0',
    inheritAttrs: false,
    props: {
      /**
       * Field id
       */
      id: {
        type: String,
        default: ''
      },
      /**
       * Use for v-model
       */
      modelValue: {
        type: [Number, String],
        default: ''
      },
      /**
       * Size of text field
       * @values medium, small
       */
      size: {
        type: String,
        default: 'medium',
        validator: value => SIZE.includes(value)
      },
      /**
       * Label for field
       */
      label: {
        type: String,
        default: ''
      },
      /**
       * Whether show disable state or not
       */
      isDisabled: {
        type: Boolean,
        default: false
      },
      /**
       * Whether show error state or not
       */
      isError: {
        type: Boolean,
        default: false
      },
    },
    emits: ['update:modelValue', 'click', 'blur', 'focus'],
    data() {
      return {
        isFocused: false
      }
    },
    computed: {
      classes() {
        let list = []

        if (this.isDisabled) {
          list.push('is-disabled')
        }

        if (this.isError) {
          list.push('is-error')
        }

        if (this.isFocused) {
          list.push('is-focused')
        }

        return list
      },
      theme() {
        let height, padding

        if(this.size === 'small') {
          height = '36px'
          padding = '0px 12px'
        } else {
          height = '44px'
          padding = '0px 20px'
        }

        return {
          height,
          padding,
        }
      }
    },
    methods: {
      handleInput(event) {
        /**
         * Triggers when value selected changes
         *
         * @event update:value
         * @type {object}
         * @arg value return event target value
         */
        this.$emit('update:modelValue', event.target.value)
      },
      handleClick(event) {
        /**
         * Triggers when dropdown clicked
         *
         * @event clicked
         * @type {object}
         * @arg event return event object
         */
        this.$emit('click', event)
      },
      handleBlur(event) {
        this.isFocused = false

        /**
         * Triggers when dropdown on blur
         *
         * @event blur
         * @type {object}
         * @arg event return event object
         */
        this.$emit('blur', event)
      },
      handleFocus(event) {
        this.isFocused = true

        /**
         * Triggers when dropdown on focus
         *
         * @event focus
         * @type {object}
         * @arg event return event object
         */
        this.$emit('focus', event)
      },
    }
  }
</script>

<style scoped>
  .label {
    display: block;
    color: #3d3d3d;
    font-family: sans-serif;
    font-size: 14px;
    line-height: 20px;
    margin-bottom: 8px;
  }

  .input {
    position: relative;
    box-sizing: border-box;
    -webkit-appearance: none;
    -moz-appearance: none;
    outline: none;
    box-shadow: none;
    letter-spacing: normal;
    color: #3d3d3d;
    font-family: sans-serif;
    font-size: 16px;
    line-height: 20px;
    width: 100%;
    height: v-bind('theme.height');
    padding: v-bind('theme.padding');
    background: #ffffff;
    /*border-radius: 8px;*/
    border: 2px solid #d6d6d6;
    transition: all .2s ease-in-out;
  }

  .input:hover {
    border-color: #2572b7;
  }

  .input::placeholder {
    color: #8d8d8d;
  }

  .is-focused {
    border-color: #2572b7;
  }

  .is-error {
    border-color: #ec0c0c;
  }

  .is-error:hover {
    border-color: #ec0c0c;
  }

  .is-disabled {
    cursor: not-allowed;
    background-color: #f2f2f2;
    border-color: #f2f2f2;
    color: #b6b6b6;
  }

  .is-disabled:hover {
    background-color: #f2f2f2;
    border-color: #f2f2f2;
    color: #b6b6b6;
  }
</style>

