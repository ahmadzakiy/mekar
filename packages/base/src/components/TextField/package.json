{
  "name": "mekar-text-field",
  "version": "0.4.0",
  "description": "component text field",
  "main": "dist/index.js",
  "scripts": {
    "clean": "rm -rf ./dist",
    "release": "npm publish"
  },
  "author": "Ahmad Zakiy <in.zakiy@gmail.com>",
  "license": "MIT",
  "devDependencies": {
    "vue": "^3.2.12"
  },
  "peerDependencies": {
    "vue": "^3.2.12"
  },
  "dependencies": {
    "vue3-styled-components": "^1.2.1"
  }
}
