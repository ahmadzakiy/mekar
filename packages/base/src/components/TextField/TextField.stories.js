import TextField from './TextField'
import { SIZE } from './type'

export default {
  title: 'Components/TextField',
  component: TextField,
  argTypes: {
    size: {
      control: 'select',
      options: SIZE
    },
    onClick: {
      action: 'click'
    },
    onFocus: {
      action: 'focus'
    },
    onBlur: {
      action: 'blur'
    },
    'onUpdate:modelValue': {
      action: 'update'
    },
  },
};

const Template = (args) => ({
  components: {
    TextField
  },
  setup() {
    return { args }
  },
  template: '<TextField v-bind="args" />',
});

export const Default = Template.bind({});
Default.args = {
  value: '',
  id: 'name',
  type: 'text',
  size: 'medium',
  placeholder: 'John Doe',
  label: 'Fill your name',
}
