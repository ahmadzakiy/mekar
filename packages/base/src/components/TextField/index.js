import TextField from './TextField.vue'

TextField.install = function(Vue) {
  Vue.component(TextField.name, TextField)
}

export default TextField
