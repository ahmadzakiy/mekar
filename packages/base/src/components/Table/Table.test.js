import { mount } from '@vue/test-utils'
import TextField from './Table'

describe('Component: Table', () => {
  let instance

  const propsObj = {
    data: [],
    columnTitle: [],
    size: 'medium',
    isStriped: false
  }

  beforeEach(() => {
    instance = mount(TextField)
  })

  afterEach(() => {
    instance.unmount()
  })

  it('Should be named Table', () => {
    expect(instance.vm.$options.name).toBe('Table')
  })

  it('Should match snapshot', async () => {
    await instance.setProps({
      columnTitle: ['name', 'age'],
      data: [
        { name: 'A', age: 1 },
        { name: 'B', age: 2 },
        { name: 'C', age: 3 },
      ]
    })

    expect(instance.element).toMatchSnapshot()
  })

  it('Should have default props', () => {
    expect(instance.props()).toEqual(propsObj)
  })

  it('Should render data correctly', async () => {
    await instance.setProps({
      columnTitle: ['name', 'age'],
      data: [
        { name: 'A', age: 1 },
        { name: 'B', age: 2 },
        { name: 'C', age: 3 },
      ]
    })

    expect(instance.findAll('th').length).toBe(2)
    expect(instance.findAll('th')[0].text()).toBe('Name')
    expect(instance.findAll('th')[1].text()).toBe('Age')

    expect(instance.findAll('td').length).toBe(6)
    expect(instance.findAll('td')[0].text()).toBe('A')
    expect(instance.findAll('td')[1].text()).toBe('1')
    expect(instance.findAll('td')[2].text()).toBe('B')
    expect(instance.findAll('td')[3].text()).toBe('2')
    expect(instance.findAll('td')[4].text()).toBe('C')
    expect(instance.findAll('td')[5].text()).toBe('3')
  })

  it('Should sort data correctly',  async () => {
    await instance.setProps({
      columnTitle: ['name', 'age'],
      data: [
        { name: 'A', age: 1 },
        { name: 'B', age: 2 },
        { name: 'C', age: 3 },
      ]
    })
    await instance.setData({
      sortOrders: {
        name: 1,
        age: 1
      }
    })

    expect(instance.findAll('td').length).toBe(6)
    expect(instance.findAll('td')[0].text()).toBe('A')
    expect(instance.findAll('td')[1].text()).toBe('1')

    await instance.findAll('th')[0].trigger('click')

    expect(instance.findAll('td')[0].text()).toBe('C')
    expect(instance.findAll('td')[1].text()).toBe('3')
  })
})
