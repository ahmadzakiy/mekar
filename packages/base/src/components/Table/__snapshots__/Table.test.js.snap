// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Table Should match snapshot 1`] = `
<table
  class="table"
>
  <thead>
    <tr>
      
      <th
        class=""
      >
        Name 
        <span
          class="arrow dsc"
        />
      </th>
      <th
        class=""
      >
        Age 
        <span
          class="arrow dsc"
        />
      </th>
      
    </tr>
  </thead>
  <tbody>
    
    <tr>
      
      <td>
        A
      </td>
      <td>
        1
      </td>
      
    </tr>
    <tr>
      
      <td>
        B
      </td>
      <td>
        2
      </td>
      
    </tr>
    <tr>
      
      <td>
        C
      </td>
      <td>
        3
      </td>
      
    </tr>
    
  </tbody>
</table>
`;
